File: packages/core/src/transport/WebViewTransport.ts
import type { Transport } from '../types.js';

export class WebViewTransport implements Transport {
  private handler: ((message: string) => void) | null = null;

  constructor() {
    if (typeof window !== 'undefined') {
      window.__rynbridge_receive = (message: string) => {
        this.handler?.(message);
      };
    }
  }

  send(message: string): void {
    if (typeof window === 'undefined') return;

    // iOS: WKWebView
    if (window.webkit?.messageHandlers?.RynBridge) {
      window.webkit.messageHandlers.RynBridge.postMessage(message);
      return;
    }

    // Android: WebView.addJavascriptInterface
    if (window.RynBridgeAndroid) {
      window.RynBridgeAndroid.postMessage(message);
      return;
    }
  }

  onMessage(handler: (message: string) => void): void {
    this.handler = handler;
  }

  dispose(): void {
    this.handler = null;
    if (typeof window !== 'undefined') {
      window.__rynbridge_receive = undefined;
    }
  }
}

File: android/core/src/main/kotlin/com/devryner/rynbridge/core/WebViewTransport.kt
package com.devryner.rynbridge.core

import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebView
import java.lang.ref.WeakReference

class WebViewTransport(webView: WebView) : Transport {

    private val webViewRef = WeakReference(webView)
    private val mainHandler = Handler(Looper.getMainLooper())
    private var messageHandler: ((String) -> Unit)? = null
    private var disposed = false

    override fun send(message: String) {
        if (disposed) return
        val escaped = message
            .replace("\\", "\\\\")
            .replace("'", "\\'")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
        mainHandler.post {
            webViewRef.get()?.evaluateJavascript(
                "window.__rynbridge_receive('$escaped')",
                null
            )
        }
    }

    override fun onMessage(handler: (String) -> Unit) {
        messageHandler = handler
    }

    override fun dispose() {
        disposed = true
        messageHandler = null
    }

    @JavascriptInterface
    fun postMessage(message: String) {
        messageHandler?.invoke(message)
    }
}
